src/lib/comms/sol-http-server-impl-microhttpd.c

```diff
@@ -0,0 +1,585 @@
+
+    if (!req) {
+        req = sol_vector_append(&server->requests);
+        SOL_NULL_CHECK(req, MHD_NO);
+
+        sol_http_param_init(&req->params);
+        req->url = url;
+        req->connection = connection;
+        *ptr = req;
+        return MHD_YES;
+    }
+
+    if (method && streq(method, "POST")) {
+        req->method = SOL_HTTP_METHOD_POST;
+        if (!req->pp)
+            req->pp = MHD_create_post_processor(connection, 1024, post_iterator, req);
+        SOL_NULL_CHECK_GOTO(req->pp, end);
+
+        MHD_post_process(req->pp, upload_data, *upload_data_size);
+        if (*upload_data_size) {
+            *upload_data_size = 0;
+            return MHD_YES;
+        }
+    } else if (method && streq(method, "GET")) {
+        req->method = SOL_HTTP_METHOD_GET;
+    } else {
+        SOL_WRN("Method %s not implemented", method);
+        status = SOL_HTTP_STATUS_NOT_IMPLEMENTED;
+        goto end;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&server->handlers, handler, i) {
+        if (!sol_str_slice_str_eq(handler->path, url))
+            continue;
+
+        MHD_get_connection_values(connection, MHD_HEADER_KIND, headers_iterator, req);
+        if (handler->last_modified && (req->if_since_modified > handler->last_modified)) {
+            status = SOL_HTTP_STATUS_NOT_MODIFIED;
+            goto end;
+        } else {
+            MHD_suspend_connection(connection);
+            ret = handler->request_cb((void *)handler->user_data, req);
+            SOL_INT_CHECK(ret, < 0, MHD_NO);
+        }
+
+        return MHD_YES;
+    }
+
+end:
+    mhd_response = MHD_create_response_from_buffer(0, NULL, MHD_RESPMEM_PERSISTENT);
+    SOL_NULL_CHECK(mhd_response, MHD_NO);
+
+    ret = MHD_queue_response(connection, status, mhd_response);
+    MHD_destroy_response(mhd_response);
+    return ret;
+}
+
+static bool
+connection_watch_cb(void *data, int fd, unsigned int flags)
+{
+    uint16_t i;
+    fd_set rs, ws, es;
+    struct sol_http_server *server = data;
+    struct http_connection *connection;
+
+    SOL_NULL_CHECK(server, false);
+
+    FD_ZERO(&rs);
+    FD_ZERO(&ws);
+    FD_ZERO(&es);
+
+    if (flags & SOL_FD_FLAGS_IN)
+        FD_SET(fd, &rs);
+    if (flags & SOL_FD_FLAGS_OUT)
+        FD_SET(fd, &ws);
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_NVAL | SOL_FD_FLAGS_HUP))
+        FD_SET(fd, &es);
+
+    /* This is far from optimized, the mainloop is given one file descriptor
+       per time while it's possible to pass all the file descriptors with events
+       to MHD at once. */
+    if (MHD_run_from_select(server->daemon, &rs, &ws, &es) == MHD_NO)
+        SOL_WRN("Something wrong happened inside microhttpd");
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_NVAL | SOL_FD_FLAGS_HUP)) {
+        SOL_VECTOR_FOREACH_IDX (&server->fds, connection, i) {
+            if (connection->fd == fd) {
+                sol_vector_del(&server->fds, i);
+                break;
+            }
+        }
+
+        return false;
```

Comment:
yes, in this case, something wrong happened with the file descriptor and the MHD was already notified, so there is nothing to do anymore. It should never happened, but .... in the worst scenario this vector is cleared in the shutdown.